lens/eval_incremental.ml

```diff
@@ -0,0 +1,217 @@
+open Value
+open Lens_utility
+module Sorted = Sorted_records
+
+let matches_change changes =
+  let is_changed ((cols_l, _cols_r), vals) =
+    let vals_l = List.map ~f:(fun (left, _) -> left) vals in
+    Phrase.Option.in_expr cols_l vals_l
+  in
+  List.map ~f:is_changed changes |> Phrase.List.fold_or_opt
+
+let delta_merge_affected lens data =
+  let sort = Value.sort lens in
+  let fun_deps = Sort.fds sort in
+  let changelist = Sorted.calculate_fd_changelist data ~fun_deps in
+  (* query relevant rows in database *)
+  let predicate = matches_change changelist in
+  let res = Value.lens_get_select_opt lens ~predicate in
+  let res =
+    Sorted.construct_cols
+      ~columns:(Value.cols_present_aliases lens)
+      ~records:res
+  in
+  (* perform relational update *)
+  let rel_merge =
+    Sorted_records.relational_merge res ~update_with:data ~fun_deps
+  in
+  Sorted.merge rel_merge (Sorted.negate res)
+
+let query_join_records lens set on =
+  let proj = Sorted.project_onto set ~columns:on in
+  let recs = Sorted.all_values proj in
+  let predicate = Phrase.Option.in_expr on recs in
+  let records = Value.lens_get_select_opt lens ~predicate in
+  Sorted.construct_cols ~columns:(Value.cols_present_aliases lens) ~records
+
+let query_project_records lens set key drop =
+  let set = Sorted.force_positive set in
+  let proj = Sorted.project_onto set ~columns:key in
+  let recs = Sorted.all_values proj in
+  let predicate = Phrase.Option.in_expr key recs in
+  let records = Value.lens_get_select_opt lens ~predicate in
+  let records =
+    Sorted.construct_cols ~columns:(Value.cols_present_aliases lens) ~records
+  in
+  Sorted.project_onto records ~columns:(List.append key drop)
+
+let lens_put_set_step lens delt (fn : Value.t -> Sorted.t -> unit) =
+  match lens with
+  | Lens _ -> fn lens delt
+  | LensDrop {lens= l; drop; key; default; _} ->
+      let relevant = query_project_records l delt [key] [drop] in
+      let delt =
+        Sorted.relational_extend delt ~key ~by:drop ~default ~data:relevant
+      in
+      fn l delt
+  | LensJoin {left; right; on; del_left; del_right; _} ->
+      let cols_simp = List.map ~f:(fun (a, _, _) -> a) on in
+      let sort1 = Value.sort left in
+      let proj1 =
+        Sorted.project_onto delt ~columns:(Sort.cols_present_aliases sort1)
+      in
+      let sort2 = Value.sort right in
+      let proj2 =
+        Sorted.project_onto delt ~columns:(Sort.cols_present_aliases sort2)
+      in
+      let delta_m0 = delta_merge_affected left proj1 in
+      let delta_n0 = delta_merge_affected right proj2 in
+      let on' = List.map ~f:(fun a -> (a, a, a)) cols_simp in
+      let delta_l =
+        Sorted.merge
```

Comment:
Why not do `let open Sorted` here, to avoid all of the prefixes?

(In fact, there's an argument for doing so before the `match`)

Reply:
I do this on purpose as I find it makes the code more readable as one can directly see which function comes from which module and applies to which type just as I usually leave the module `List` etc. It also works better with autocompletion, as when looking for a specific function typing in `Sorted.` brings up a relevant list: 

![image](https://user-images.githubusercontent.com/6114830/61140900-446c3c00-a4c4-11e9-8679-e1e1ba259c58.png)

Reply:
OK, matter of taste -- fine by me.